dace/transformation/dataflow/tasklet_fusion.py

```diff
@@ -0,0 +1,226 @@
+# Copyright 2019-2021 ETH Zurich and the DaCe authors. All rights reserved.
+""" Contains classes that fuse Tasklets """
+
+from typing import Any, Dict, Set
+import ast
+import dace
+import re
+from dace import dtypes, registry
+from dace.sdfg import nodes
+from dace.sdfg import utils as sdutil
+from dace.transformation import transformation as pm
+
+
+class ConnectorRenamer(ast.NodeTransformer):
+    """ Renames connector names in Tasklet code.
+    """
+    def __init__(self, repl_dict: Dict[str, str]) -> None:
+        """ Initializes AST transformer.
+            :param repl_dict: Replacement dictionary.
+        """
+        self.repl_dict = repl_dict
+
+    def visit_Name(self, node: ast.Name) -> Any:
+        # Rename connector
+        if node.id in self.repl_dict:
+            node.id = self.repl_dict[node.id]
+        return self.generic_visit(node)
+
+
+class PythonRHSExtractor(ast.NodeVisitor):
+    """ Extracts assignments' RHS in Tasklet code.
+    """
+    def __init__(self):
+        self.assignments = set()
+
+    def visit_Assign(self, node: ast.Assign):
+        if isinstance(node.targets[0], ast.Name):
+            self.assignments.add(node.targets[0].id)
+
+    def visit_AugAssign(self, node: ast.AugAssign):
+        if isinstance(node.targets[0], ast.Name):
+            self.assignments.add(node.targets[0].id)
+
+
+@registry.autoregister_params(singlestate=True, strict=False)
+class SimpleTaskletFusion(pm.Transformation):
+    """ Fuses two connected Tasklets.
+        It is recommended that this transformation is used on Tasklets that
+        contain only simple assignments.
+    """
+
+    _t1 = pm.PatternNode(nodes.Tasklet)
+    _t2 = pm.PatternNode(nodes.Tasklet)
+
+    @staticmethod
+    def expressions():
+        return [
+            sdutil.node_path_graph(SimpleTaskletFusion._t1,
+                                   SimpleTaskletFusion._t2)
+        ]
+
+    @staticmethod
+    def can_be_applied(graph: dace.SDFGState,
+                       candidate: Dict[pm.PatternNode, int],
+                       expr_index: int,
+                       sdfg: dace.SDFG,
+                       strict: bool = False):
+
+        t1 = graph.node(candidate[SimpleTaskletFusion._t1])
+        t2 = graph.node(candidate[SimpleTaskletFusion._t2])
+
+        # Tasklets must be of the same language
+        if t1.language != t2.language:
+            return False
+
+        # Avoid circles
+        t1_dst = set()
+        for e in graph.out_edges(t1):
+            t1_dst.add(e.dst)
+        t2_src = set()
+        for e in graph.in_edges(t2):
+            t2_src.add(e.src)
+        if len(t1_dst.intersection(t2_src)):
+            return False
```

Comment:
Isn't this guaranteed to never happen by DaCe semantics?

Reply:
The transformation always fuses the second tasklet (`t2`) to the first one (`t1`). The check is there to enforce a certain ordering of fusions so that a circle is not produced due to this design. A simple example is the following:

Assume that the tasklet subgraph contains the edges:
- `t1 -> t2`
- `t1 -> t3`
- `t3 -> t2`

If we fuse `t2` to `t1`, then the `t3 -> t2` edge will become `t3 -> t1`, producing a cycle. The subgraph can of course be fused. We just have to fuse first `t3` to `t1`.